src/util/amount.rs

```diff
@@ -0,0 +1,367 @@
+
+impl<P: Precision> Amount<P> {
+    pub fn from_inner(amount: Inner) -> Amount<P> {
+        Amount(amount, Default::default())
+    }
+
+    pub fn from_sat(satoshis: i64) -> Amount<P> {
+        assert!(P::scale() >= 0);
+        Amount::from_inner(rescale_inner(satoshis, P::scale()))
+    }
+
+    pub fn from_btc<T: IntoBtc<P>>(btc: T) -> Amount<P> {
+        btc.into_btc()
+    }
+
+    pub fn zero() -> Amount<P> {
+        Amount::from_inner(0)
+    }
+
+    pub fn one() -> Amount<P> {
+        Amount::from_inner(1)
+    }
+
+    pub fn as_inner(&self) -> Inner {
+        self.0
+    }
+
+    pub fn into_inner(self) -> Inner {
+        self.0
+    }
+
+    pub fn float_denomination(&self, denom: Denomination) -> f64 {
+        let exp = -P::scale() - denom as i32;
+        (self.0 as f64) * 10f64.powi(exp)
+    }
+
+    pub fn fmt_denomination(&self, f: &mut fmt::Write, denom: Denomination) -> fmt::Result {
+        write!(f, "{} {}", self.float_denomination(denom.clone()), denom)
+    }
+
+    pub fn to_denominated_string(&self, denom: Denomination) -> String {
+        let mut buf = String::new();
+        self.fmt_denomination(&mut buf, denom).unwrap();
+        buf
+    }
+
+    pub fn rescaled<O: Precision>(&self) -> Amount<O> {
+        let amt = rescale_inner(self.0, O::scale() - P::scale());
+        Amount::from_inner(amt)
+    }
+
+    pub fn into_rescaled<O: Precision>(self) -> Amount<O> {
+        let amt = rescale_inner(self.0, O::scale() - P::scale());
+        Amount::from_inner(amt)
+    }
+}
+
+impl<P: Precision> fmt::Debug for Amount<P> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Amount({}, presision={})", self.0, P::scale())
+    }
+}
+
+impl fmt::Display for Amount {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.fmt_denomination(f, Denomination::Bitcoin)
+    }
+}
+
+impl<P: Precision> ops::Add for Amount<P> {
+    type Output = Amount<P>;
+
+    fn add(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 + rhs.0)
+    }
+}
+
+impl<P: Precision> ops::Sub for Amount<P> {
+    type Output = Amount<P>;
+
+    fn sub(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 - rhs.0)
+    }
+}
+
+impl<P: Precision> ops::Mul<i64> for Amount<P> {
+    type Output = Amount<P>;
+
+    fn mul(self, rhs: i64) -> Self::Output {
+        Amount::from_inner(self.0 * rhs)
+    }
+}
+
+impl<P: Precision> ops::Div for Amount<P> {
+    type Output = f64;
+
+    fn div(self, rhs: Amount<P>) -> Self::Output {
+        self.0 as f64 / rhs.0 as f64
+    }
+}
+
+impl<P: Precision> ops::Div<f64> for Amount<P> {
+    type Output = Amount<P>;
+
+    fn div(self, rhs: f64) -> Self::Output {
+        Amount::from_inner((self.0 as f64 / rhs as f64) as i64)
+    }
+}
+
+impl PartialEq for Amount {
+    fn eq(&self, other: &Amount) -> bool {
+        PartialEq::eq(&self.0, &other.0)
+    }
+}
+impl Eq for Amount {}
+
+impl PartialOrd for Amount {
```

Comment:
As long as two `Amount`s have the same `P` can't you just derive `(Partial)Ord` and `(Partial)Eq`?

Reply:
(And I think `derive` enforces that they need the same `P`)

Reply:
Yeah I wasn't sure about the Eq and Ord of those empty struct types. But I'm pretty sure you're right.